File: CorePushRichSample/CorePushSample/NotificationHistoryViewController.m
//
//  NotificationHistoryViewController.m
//  CorePushSample
//
//  Copyright (c) 2012年 株式会社ブレスサービス. All rights reserved.
//

#import "NotificationHistoryViewController.h"
#import <CorePush/CorePushNotificationHistoryModel.h>

@interface NotificationHistoryViewController ()

@end

@implementation NotificationHistoryViewController

@synthesize tableView = tableView_;


- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        self.tabBarItem.title = @"通知履歴";
        self.tabBarItem.image = [UIImage imageNamed:@"history.png"];
    }
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    self.title = @"通知履歴";
    self.tableView.delegate = self;
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    [[CorePushNotificationHistoryManager shared] setDelegate:self];
    
    // 通知履歴一覧を取得する
    [UIApplication sharedApplication].networkActivityIndicatorVisible = YES;
    [[CorePushNotificationHistoryManager shared] requestNotificationHistory];
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return [[[CorePushNotificationHistoryManager shared] notificationHistoryModelArray] count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView
         cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    static NSString *CellIdentifier = @"Cell";
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier];
    if (cell == nil) {
        cell = [[[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:CellIdentifier] autorelease];
    }

    CorePushNotificationHistoryModel* historyModel = [[[CorePushNotificationHistoryManager shared] notificationHistoryModelArray] objectAtIndex:indexPath.row];
    cell.textLabel.text = historyModel.message;
    
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
 
    CorePushNotificationHistoryModel* historyModel = [[[CorePushNotificationHistoryManager shared] notificationHistoryModelArray] objectAtIndex:indexPath.row];
    NSString* historyId = historyModel.historyId;
    
    //タップされた場合、該当する通知メッセージを既読に設定する。
    [[CorePushNotificationHistoryManager shared] setRead:historyId];
    
    //未読数をタブに設定する。
    [self setUnreadNumber];
    
    [UIApplication sharedApplication].networkActivityIndicatorVisible = NO;
    [tableView_ reloadData];
}

//通知履歴の取得成功
- (void)notificationHistoryManagerSuccess {
    [self setUnreadNumber];
    
    [UIApplication sharedApplication].networkActivityIndicatorVisible = NO;
    [tableView_ reloadData];
}

//通知履歴の取得失敗
- (void)notificationHistoryManagerFail {
    [UIApplication sharedApplication].networkActivityIndicatorVisible = NO;
}

//未読数をタブに表示する。
- (void)setUnreadNumber {
    int unreadNumber = [[CorePushNotificationHistoryManager shared] getUnreadNumber];
    if (unreadNumber > 0) {
        self.tabBarItem.badgeValue = [NSString stringWithFormat:@"%d", unreadNumber];
        
    } else {
        self.tabBarItem.badgeValue = nil;
    }
}

- (void)dealloc {
    [tableView_ release], tableView_ = nil;
    [super dealloc];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
}

@end

File: CorePushRichSample_Swift/CorePushSample_Swift/NotificationHistoryViewController.swift
//
//  NotificationHistoryViewController.swift
//  CorePushSample_Swift
//
//  Copyright (c) 2016 株式会社ブレスサービス. All rights reserved.
//

import UIKit

class NotificationHistoryViewController: UIViewController {
    
    @IBOutlet weak var tableView: UITableView!
    
    static let CellIdentifier = "Cell"
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.title = "通知履歴";
        self.tableView.delegate = self
        self.tableView.dataSource = self
        
        self.tableView.register(UITableViewCell.self, forCellReuseIdentifier: NotificationHistoryViewController.CellIdentifier)
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        let historyManager = CorePushNotificationHistoryManager.shared() as! CorePushNotificationHistoryManager
        historyManager.delegate = self
        
        // 通知履歴一覧を取得する
        UIApplication.shared.isNetworkActivityIndicatorVisible = true
        historyManager.requestNotificationHistory()
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
    }
    
    func setUnreadNumber() {
        let historyManager = CorePushNotificationHistoryManager.shared() as! CorePushNotificationHistoryManager
        let unreadNumber = historyManager.getUnreadNumber()
        if unreadNumber > 0 {
            self.navigationController?.tabBarItem.badgeValue = "\(unreadNumber)"
        } else {
            self.navigationController?.tabBarItem.badgeValue = nil;
        }
    }
}

extension NotificationHistoryViewController: UITableViewDelegate {
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        
        tableView.deselectRow(at: indexPath, animated: true)
        
        let historyManager = CorePushNotificationHistoryManager.shared() as! CorePushNotificationHistoryManager
        let historyModelArray = historyManager.notificationHistoryModelArray as NSMutableArray
        let historyModel = historyModelArray.object(at: indexPath.row) as! CorePushNotificationHistoryModel
        let historyId = historyModel.historyId
        
        //タップされた場合、該当する通知メッセージを既読に設定する。
        historyManager.setRead(historyId)
        
        //未読数をタブに設定する。
        setUnreadNumber()
        
        UIApplication.shared.isNetworkActivityIndicatorVisible = false
        tableView.reloadData()
    }
}

extension NotificationHistoryViewController: UITableViewDataSource {
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        let historyManager = CorePushNotificationHistoryManager.shared() as! CorePushNotificationHistoryManager
        let historyModelArray = historyManager.notificationHistoryModelArray as NSMutableArray
        return historyModelArray.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {

        let cell = tableView.dequeueReusableCell(withIdentifier: NotificationHistoryViewController.CellIdentifier, for: indexPath)
      
        let historyManager = CorePushNotificationHistoryManager.shared() as! CorePushNotificationHistoryManager
        let historyModelArray = historyManager.notificationHistoryModelArray as NSMutableArray
        let historyModel = historyModelArray.object(at: indexPath.row) as! CorePushNotificationHistoryModel
        
        cell.textLabel?.text = historyModel.message
        
        return cell
    }
}

extension NotificationHistoryViewController: CorePushNotificationHistoryManagerDelegate {
    
    //通知履歴の取得成功
    func notificationHistoryManagerSuccess() {
        setUnreadNumber()
        UIApplication.shared.isNetworkActivityIndicatorVisible = false
        tableView.reloadData()
    }
    
    //通知履歴の取得失敗
    func notificationHistoryManagerFail() {
        UIApplication.shared.isNetworkActivityIndicatorVisible = false
    }
}
